File: non_membership_testing/poseidon_hash.py
from poseidon_constants import POSEIDON_C
from poseidon_constants import POSEIDON_M
import time

p = 21888242871839275222246405745257275088548364400416034343698204186575808495617

def exp5(in1):
	in2 = in1*in1 % p
	in4 = in2*in2 % p 
	in5 = in4*in1 % p

	return in5

def ark(state, C, r):
	for i in range(0, len(state)):
		state[i] = state[i] + C[i+r] % p
	
	return state

def mix(state, M):
	lc = 0
	output = []

	for i in range(0, len(state)):
		lc = 0
		for j in range(0, len(state)):
			lc = lc + M[i][j]*state[j] % p
		output.append(lc)

	return output

def sbox(nRoundsF, nRoundsP, state, i):
	if ((i < nRoundsF/2) or (i >= nRoundsF/2 + nRoundsP)):
		for j in range(0, len(state)):
			state[j] = exp5(state[j])
	else:
		state[0] = exp5(state[0])

	return state




def poseidon_hash(input):
    
    nInputs = len(input)
    N_ROUNDS_P = [56, 57, 56, 60, 60, 63, 64, 63]
    t = nInputs + 1
    nRoundsF = 8
    nRoundsP = N_ROUNDS_P[t - 2]

    C = POSEIDON_C(t)
    M = POSEIDON_M(t)

    state = [0]
    for item in input:
    	state.append(item)

    for i in range(0, nRoundsF + nRoundsP):
    	# print(i," ", state[0])
    	# print(i," in ",hex(state[0]),state[0])
    	state = ark(state, C, i*t)
    	# print(i," ark ",hex(state[0]), state[0])
    	state = sbox(nRoundsF, nRoundsP, state, i)
    	# print(i," sbox ",state[0])
    	state = mix(state, M)
    	# print(i," mix ",state[0])

    output = state[0] % p
    return output

# a = poseidon_hash([3993726729728456562288376524736781253942224580193245007545935011759967092819, 20532736310324258262994538691311925383259813537971010774634242964482628474019]) 
# print(a)
# 0x295f05722342fa8f3598a3fef7b69add734ea2ab1f64c9a6974d955b113a696f

File: readable_code/PoseidonHash.java
public class PoseidonHash {

	// constant used in Poseidon Hash (8 input, bn254) 
	// The same as the one in circomlib: https://github.com/iden3/circomlib/blob/master/circuits/poseidon_constants.circom
	public static string[][] POSEIDON_C = {{
		// the actual strings are omitted for space
	}};

	// 2-input Posedion Hash, each input is F_p element 
// The same as the one in circomlib: https://github.com/iden3/circomlib/blob/master/circuits/poseidon.circom 
 
public static F_p poseidon_hash(F_p[] input) { 
   
  if (input.length != 2) { throw new IllegalArgumentException("This method only accepts 2 Fp elements as inputs"); } 
   
  int nInput = 2; 
  int t = nInput + 1; 
  int[] N_ROUND_P = {56, 57, 56, 60, 60, 63, 64, 63}; 
   
  F_p[] C = F_p(POSEIDON_C[t - 2]); 
  F_p[][] M = new F_p[t][t]; 
   
  for (int i = 0; i < t; i++) { 
    M[i] = F_p(POSEIDON_M[t - 2][i]); 
  } 
   
  int nRoundsF = 8; 
  int nRoundsP = N_ROUND_P[t - 2]; 
   
  F_p[] state = new F_p[t]; 
  state[0] = F_p("0"); 
   
  for (int i = 0; i < nInput; i++) { 
    state[i + 1] = input[i]; 
  } 
   
  // This is computation in each round, do ark, abox and then mix 
  // Here we merge different rounds in the same sbox function! 
  for (int i = 0; i < nRoundsF + nRoundsP; i++) { 
    state = ark(state, C, i * t); 
    state = sbox(nRoundsF, nRoundsP, state, i); 
    state = mix(state, M); 
  } 
   
  F_p output = state[0]; 
   
  return output; 
} 
 
// 8-input Posedion Hash, each input is F_p element 
// The same as the one in circomlib: https://github.com/iden3/circomlib/blob/master/circuits/poseidon.circom 
 
public static F_p poseidon_hash_8(F_p[] input) { 
   
  if (input.length != 8) { throw new IllegalArgumentException("This method only accepts 2 Fp elements as inputs"); } 
   
  int nInput = 8; 
  int t = nInput + 1; 
  int[] N_ROUND_P = {56, 57, 56, 60, 60, 63, 64, 63}; 
   
  F_p[] C = F_p(POSEIDON_C[t - 2]); 
  F_p[][] M = new F_p[t][t]; 
   
  for (int i = 0; i < t; i++) { 
    M[i] = F_p(POSEIDON_M[t - 2][i]); 
  } 
   
  int nRoundsF = 8; 
  int nRoundsP = N_ROUND_P[t - 2]; 
   
  F_p[] state = new F_p[t]; 
  state[0] = F_p("0"); 
   
  for (int i = 0; i < nInput; i++) { 
    state[i + 1] = input[i]; 
  } 
   
  // This is computation in each round, do ark, abox and then mix 
  // Here we merge different rounds in the same sbox function! 
  for (int i = 0; i < nRoundsF + nRoundsP; i++) { 
    state = ark(state, C, i * t); 
    state = sbox(nRoundsF, nRoundsP, state, i); 
    state = mix(state, M); 
  } 
   
  F_p output = state[0]; 
   
  return output; 
} 
 
// Exp5 function in Poseidon Hash, return in**5 
 
public static F_p exp5(F_p in) { 
  F_p in2; 
  F_p in4; 
  F_p out; 
   
  in2 = in * in; 
  in4 = in2 * in2; 
  out = in4 * in; 
   
  return out; 
} 
 
// Ark function in Poseidon Hash, return State+Matrix 
 
public static F_p[] ark(F_p[] state, F_p[] C, int r) { 
  for (int i = 0; i < state.length; i++) { 
    state[i] = state[i] + C[i + r]; 
  } 
   
  return state; 
} 
 
// Mix function in Poseidon Hash 
 
public static F_p[] mix(F_p[] state, F_p[][] M) { 
  F_p lc; 
  F_p[] output = new F_p[state.length]; 
   
  for (int i = 0; i < state.length; i++) { 
    lc = 0; 
    for (int j = 0; j < state.length; j++) { 
      lc = lc + M[i][j] * state[j]; 
    } 
    output[i] = lc; 
  } 
   
  return output; 
} 
 
// Sbox function in Poseidon Hash 
// Note that we do different operations in different rounds! (merge if else here) 
 
public static F_p[] sbox(int nRoundsF, int nRoundsP, F_p[] state, int i) { 
  if ((i < nRoundsF / 2) || (i >= nRoundsF / 2 + nRoundsP)) { 
    for (int j = 0; j < state.length; j++) { 
      state[j] = exp5(state[j]); 
    } 
  } else { 
    state[0] = exp5(state[0]); 
  } 
   
  return state; 
} 


}
